rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+- Connect to a relay (`directly`/`only-unknown`) or reuse existing connection and `SEND`.
+- Connect to a proxy or pick one from a pool (excluding the same address as the destination relay for `force-hop` mode) and
+  * Request a new relay connection.
+  * `FORWARD` message to a proxy.
+
+For group traffic a client will have to utilize a pool of proxies since many recipients can be sitting on the same relay, bottlenecking delivery on the user cluster.
+An agent can spread delivery to the same relay across multiple proxies, minimizing per-server queue length.
+This will also help with protecting the group participation information from servers and makes a case for using multiple SMP providers.
+An alternative server-side batching command makes membership correlation trivial for malicious servers.
+
+### Proxy server
+
+Running a proxy service adds another potentially unbounded source of load to a server.
+Previously, a client could flood relay with queue creation requests or pump messages through it, generating local load.
+With proxy enabled, a client may now ask a server to open connections to arbitrary hosts/ports, which may affect external services, not unlike open access proxies. An SMP proxy, by the virtue of being app-specific, can enforce more guardrails, limiting problems from malicious clients.
+
+- Failed `server` requests may incur a penalty/backpressure to protect external services and proxy.
+- Forwarding is blocked until a reply comes back to protect proxy.
+- Minimal client response time at ~2 seconds to protect client.
+- Forwarding requests are pulled from client queues and handled one by one to protect relay.
+- Fair client scheduling towards a relay to protect clients from each other.
+
+Each proxy-to-relay (P2R) connection aggregates traffic from multiple clients and protected by a
+
+A proxy runs a barebones SMP protocol towards each active relay, fronted by an `MVar` which, unlike other concurrency primitives, guarantee fairness. With it, each client "forwarding" worker takes a unit of work and tries to submit it to a relay worker.
+
+A client should not be able to inspect/probe relay connections and distinguish a fresh connection from an existing one. Thus each client gets a `server_id` map with ids distinct from the proxy internal relay map. This also precludes accidental coupling between implementation and protocol format with its (quite severe) limits.
+
+### Relay server
+
+As with notification subscriber "clients" a proxy is just an another client with a special set of commands for setup and forwarding.
+The usual SMP precautions apply.
+
+## Failure modes
+
+### Delivery problems
+
+A delivery latency may increase due to:
+- Slow client/proxy handshake.
+- Slow proxy/relay handshake for new servers.
+- Long time waiting for its turn to send due to lots of active clients.
+- Long time to deliver to relay due to overloaded server.
+This can be mitigated by proxy pool.
```

Comment:
Illogical. If the server is overloaded, adding more proxies to talk to it won't reduce the load, on the opposite.